design/baremetal-operator/firmware-image-update.md

```diff
@@ -0,0 +1,195 @@
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_1090120019
+      version: 24.34.1002
+    - componentID: enp59s0f0np0_b8_ce_f6_8d_c8_62
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_0000000652
+      version: 22.35.1012
+  lastUpdated: 2022-12-27T01:12:17Z
+  conditions: ...
+```
+
+### HostFirmwareSettings Controller Changes
+
+The proposal is to extend the `HostFirmwareSettings` controller which was introduced in the `bulk-set-bios-config` proposal. The controller will be extended to track the `HostFirmwareImage` CRD. The HostFirmwareImage.spec is the desired firmware version and the HostFirmwareImage.status contain the actual firmware. Please note that the correct proposal addresses Nvidia NICs but can be extended to other vendors too as well.
+
+The NIC firmware version exists today in the data.Extra.Network struct. We will also extend `HardwareData` with the NIC firmware by adding firmware attribute to the details.NIC. BIOS version already exists in `HardwareData` but other devices which do not firmware version is not exposed will required to provide mechanism to retrieve actual firmware and store it in the `HardwareData`.
+
+The reconcile loop of the `HostFirmwareSettings` controller will do the following:
+
+- Validate the `HostFirmwareImage` attributes (e.g., vendor name, component …). We will try to validate at match as we can. If valid set condition to FirmwareImageValid to true.
+- Fetch the corresponding HardwareData and get the Nvidia NIC firmware. The Nvidia firmware contains firmware version and componentFlavor "24.34.1002 (MT_0000000540)" other NIC vendors can extend the parsing according to their NIC firmware string.
+- Fetch `HostFirmwareImage` and append missing componentID and its corresponding vendor, component and componentFlavor to the firmwares field.
+- Compare the actual componentID firmware version with the desired version
+- If not equal set `FirmwareImageChangeDetected` condition to true.
+
+The reconcile loop in the BMO:
+
+- Fetch the `HostFirmwareImage`.
+- Check if the `FirmwareImageChangeDetected` and `FirmwareImageValid` set true.
+- Add the prepareData.ActualFirmwareImage and prepareData.TargetFirmwareImage.
+- Call the prov.Prepare with the prepareData.
+- The ironic provisioner will translate the TargetFirmwareImage to manual cleaning steps.
+- On success update the `HardwareData` NIC firmware.
+
+Note: The controller will not reconcile a vendor component in the following cases:
+
+- Vendor, component or componentID in the spec section do not exist in the bare metal.
+- Actual firmware version exists in the status but no matching component or componentFlavor in the spec.
+
+### Implementation Details/Notes/Constraints
+
+The `HostFirmwareImage` spec section will be translated to ironic manual cleaning steps in the ironic provisioner as following:
+
+```shell
+baremetal node clean control-0 --clean-steps nvidia_nic_firmware_image.json

```

Comment:
Again, this is a vendor-specific step that would have to be hard-coded into the baremetal-operator.